File: 04_quicksort/typescript/04_DC.ts

const list = [1,2,3,13,4,5,6];

/**
 * 
 * @param arr number[]
 * @returns number
 */
function sum(arr: number[]): number {
  if(arr.length == 0) return 0
  return arr[0] + sum(arr.slice(1))
}

console.log("Sum: " + sum(list));

/**
 * 
 * @param arr any[]
 * @returns number
 */
function len<T>(arr: T[]): number {
  let count = 0;
  for(let i in arr) count++;
  if(count === 0) return 0
  return 1 + len(arr.slice(1));
}

console.log("Length: " + len(list));

function max(arr: number[]): number {
  let count = 0;
  for(let i in arr) count++;
  if(count == 2) return arr[0] > arr[1] ? arr[0] : arr[1];
  let subMax = max(arr.slice(1));
  return arr[0] > subMax ? arr[0] : subMax;
}

console.log("Max: " + max(list));

File: 04_quicksort/python/04_DC.py
list = [1,6,3,10,8,3,52,4,28,33,41,88]

# Recursive function to sum
def my_sum(arr):
  if arr == []:
    return 0
  return arr[0] + my_sum(arr[1:])

print("Sum: " + str(my_sum(list)))

# Recursive function to get the length
def my_len(arr):
  if arr == []:
    return 0
  return 1 + my_len(arr[1:])

print("Length: " + str(my_len(list)))

# Recursive function to get the max
def my_max(arr):
  if my_len(arr) == 2:
    return arr[0] if arr[0] > arr[1] else arr[1]
  sub_max = my_max(arr[1:])
  return arr[0] if arr[0] > sub_max else sub_max

print(my_max(list))